Components/Sensors/BatteryTask.cpp

```diff
@@ -0,0 +1,193 @@
+#include "GPIO.hpp"
+#include "TelemetryMessage.hpp"
+
+/* Macros --------------------------------------------------------------------*/
+
+/* Structs -------------------------------------------------------------------*/
+
+/* Constants -----------------------------------------------------------------*/
+
+/* Variables -----------------------------------------------------------------*/
+
+/* Prototypes ----------------------------------------------------------------*/
+
+/* Functions -----------------------------------------------------------------*/
+/**
+ * @brief Default constructor, sets and sets up storage for member variables
+ */
+BatteryTask::BatteryTask() : Task(BATTERY_TASK_QUEUE_DEPTH_OBJS)
+{
+    data = (BatteryData*)soar_malloc(sizeof(BatteryData));
+}
+
+/**
+ * @brief Creates a task for the FreeRTOS Scheduler
+ */
+void BatteryTask::InitTask()
+{
+    // Make sure the task is not already initialized
+    SOAR_ASSERT(rtTaskHandle == nullptr, "Cannot initialize battery task twice");
+
+    // Start the task
+    BaseType_t rtValue =
+        xTaskCreate((TaskFunction_t)BatteryTask::RunTask,
+            (const char*)"PTTask",
+            (uint16_t)BATTERY_TASK_STACK_DEPTH_WORDS,
+            (void*)this,
+            (UBaseType_t)BATTERY_TASK_RTOS_PRIORITY,
+            (TaskHandle_t*)&rtTaskHandle);
+
+    //Ensure creation succeded
+    SOAR_ASSERT(rtValue == pdPASS, "BatteryTask::InitTask() - xTaskCreate() failed");
+}
+
+/**
+ * @brief BatteryTask run loop
+ * @param pvParams Currently unused task context
+ */
+void BatteryTask::Run(void * pvParams)
+{
+    while (1) {
+        Command cm;
+
+        //Wait forever for a command
+        qEvtQueue->ReceiveWait(cm);
+
+        //Process the command
+        HandleCommand(cm);
+    }
+}
+
+/**
+ * @brief Handles a command
+ * @param cm Command reference to handle
+ */
+void BatteryTask::HandleCommand(Command& cm)
+{
+    //TODO: Since this task will stall for a few milliseconds, we may need a way to eat the whole queue (combine similar eg. REQUEST commands and eat to WDG command etc)
+    //TODO: Maybe a HandleEvtQueue instead that takes in the whole queue and eats the whole thing in order of non-blocking to blocking
+
+    //Switch for the GLOBAL_COMMAND
+    switch (cm.GetCommand()) {
+    case REQUEST_COMMAND: {
+        HandleRequestCommand(cm.GetTaskCommand());
+    }
+    case TASK_SPECIFIC_COMMAND: {
+        break;
+    }
+    default:
+        SOAR_PRINT("BatteryTask - Received Unsupported Command {%d}\n", cm.GetCommand());
+        break;
+    }
+
+    //No matter what we happens, we must reset allocated data
+    cm.Reset();
+}
+
+/**
+ * @brief Handles a Request Command
+ * @param taskCommand The command to handle
+ */
+void BatteryTask::HandleRequestCommand(uint16_t taskCommand)
+{
+    //Switch for task specific command within DATA_COMMAND
+    switch (taskCommand) {
+    case BATTERY_REQUEST_NEW_SAMPLE:
+        SampleBatteryVoltage();
+        GetPowerState();
```

Comment:
GetPowerState() return param goes unused here. You could add the Proto::Battery::power_source enum to BatteryData in Data.h and use GetPowerState() to only update that variable.

Reply:
I don't think we should add it to Data.h / set it in GetPowerState() because then the contents of BatteryData::p_source can get desynced with the real state.

So if anyone else in the future doing something for battery task decides the stored p_source is the 'source of truth' and doesn't use GetPowerState they'll get confused and have to debug it.

Reply:
Ah ok yes that sounds good! Please remove the function call